matscipy/dislocation.py

```diff
@@ -2499,141 +2705,163 @@ def plot_unit_cell(self, ms=250, ax=None):
         ax.set_xlabel(r"$\AA$")
         ax.set_ylabel(r"$\AA$")
 
-    def self_consistent_displacements(self, solver, bulk_positions, center, 
+    def self_consistent_displacements(self, solvers, bulk_positions, core_positions, 
                                       tol=1e-6, max_iter=100, verbose=True):
+        '''
+        Compute dislocation displacements self-consistently, with max_iter capping the number of iterations
+        Each dislocation core uses a separate solver, which computes the displacements associated with positions 
+        relative to that core (i.e. that the core is treated as lying at [0, 0, 0])
+
+        Parameters
+        ----------
+        solvers: list
+            List of functions, solvers[i] computes the displacement field for the dislocation
+            defined by core_positions[i, :]
+        bulk_positions: np.array
+            (N, 3) array of positions of atoms in the bulk configuration
+        core_positions: np.array
+            Positions of each dislocation core (shape = (ncores, 3))
+        tol: float
+            Displacement tolerance for the convergence of the self-consistent loop
+        max_iter: int
+            Maximum number of iterations of the self-consistent cycle to run before throwing 
+            an exception
+        verbose: bool
+            Enable/Disable printing progress of the self-consistent cycle each iteration
+
+        Returns
+        -------
+        displacements: np.array
+        shape (N, 3) array of displacements to apply to bulk_positions to form
+        the dislocation structure
+
+        Raises a RuntimeError if max_iter is reached without convergence
+        '''
+        if len(core_positions.shape) == 1:
+            core_positions = core_positions[np.newaxis, :]
+
+        ncores = core_positions.shape[0]
+
+        disp1 = np.zeros_like(bulk_positions)
+        for i in range(ncores):
+            disp1 += solvers[i](bulk_positions - core_positions[i, :]).real
         
-        disp1 = np.real(solver(bulk_positions - center))
         if max_iter == 0:
             return disp1
         
         # Self-consistent calculation
         res = np.inf
-        i = 0
-        while res > tol:
+        for i in range(max_iter):
             disloc_positions = bulk_positions + disp1
-            disp2 = np.real(solver(disloc_positions - center))
+            
```

Comment:
add a unit test which checks to higher precision whether two approaches to SCF give same result

Reply:
Unittest checks all dissociated dislocations to make sure difference is less than 1% of alat for a cyl of r=40.0 with partial_distance=5.0